### PR TITLE
Pull Request for Issue1679: 3) Amptek Continuous Data API

### PR DESCRIPTION
**Pull request by davidChevrier**
Changes have been made to update the continuous data API in AMDetector and implement it in CLSAmptekSDD123DetectorNew.

There are some hanging debugs and temporary changes that I have elected to leave it ... hopefully these will help guide the person or people applying these tools in the continuous 1D scan controller.

### CODE REVIEW

File: source/acquaman/AMAgnosticDataAPI.cpp [hunk 1]
```diff
@@ -199,13 +205,15 @@ AMAgnosticDataAPIAxisValueFinishedMessage::AMAgnosticDataAPIAxisValueFinishedMes
 
 AMAgnosticDataAPIDataAvailableMessage::~AMAgnosticDataAPIDataAvailableMessage(){}
 
-AMAgnosticDataAPIDataAvailableMessage::AMAgnosticDataAPIDataAvailableMessage(const QString &uniqueID, QList<double> detectorData, QList<int> detectorDimensionalitySizes, QStringList detectorDimensionalityNames, QStringList detectorDimensionalityUnits) :
+AMAgnosticDataAPIDataAvailableMessage::AMAgnosticDataAPIDataAvailableMessage(const QString &uniqueID, QList<double> detectorData, QList<int> detectorDimensionalitySizes, QStringList detectorDimensionalityNames, QStringList detectorDimensionalityUnits, bool detectorUsesAMDS) :
```
Comment (dretrex) [inline source/acquaman/AMAgnosticDataAPI.cpp]: Technically, these should be `const QList<double> &` lists.

File: source/acquaman/AMAgnosticDataAPI.cpp [hunk 2]
```diff
@@ -258,6 +266,25 @@ QStringList AMAgnosticDataAPIDataAvailableMessage::detectorDimensionalityUnits()
 	return retVal;
 }
 
+bool AMAgnosticDataAPIDataAvailableMessage::detectorUsesAMDS() const{
+	QVariant detectorUsesAMDSVariant = jsonData_.value(AMAgnosticDataAPIDefinitions::nameFromInputType(AMAgnosticDataAPIDefinitions::DetectorUsesAMDS));
+	double retVal;
```
Comment (dretrex) [inline source/acquaman/AMAgnosticDataAPI.cpp]: `double` should be `bool`, no?

File: source/actions3/actions/AMDetectorReadAction.cpp [hunk 3]
```diff
@@ -90,44 +85,55 @@ void AMDetectorReadAction::onDetectorNewValuesAvailable(){
...
+				for(int x = 0; x < totalPoints; x++)
+					detectorData.append(detectorDataPointer[x]);
+			}
 		}
-		else{
-			int totalPoints = AMnDIndex(detector_->rank(), AMnDIndex::DoInit, 0).totalPointsTo(detector_->size())-1;
-			for(int x = 0; x < totalPoints; x++)
-				detectorData.append(detectorDataPointer[x]);
+		else if(detector_->readMode() == AMDetectorDefinitions::ContinuousRead){
+			dimensionSizes.append(0);
```
Comment (dretrex) [inline source/actions3/actions/AMDetectorReadAction.cpp]: Are these placeholders?

File: source/actions3/actions/AMDetectorReadAction.cpp [hunk 4]
```diff
@@ -90,44 +85,55 @@ void AMDetectorReadAction::onDetectorNewValuesAvailable(){
...
+			dimensionNames.append("0");
+			dimensionUnits.append("0");
+
+			detectorData.append(0);
+
+			AMAgnosticDataAPIDataAvailableMessage dataAvailableMessage(detector_->name(), detectorData, dimensionSizes, dimensionNames, dimensionUnits, true);
+			intptr_t continuousDataPointer = (intptr_t)(detector_->lastContinuousData(1));
```
Comment (dretrex) [inline source/actions3/actions/AMDetectorReadAction.cpp]: Hard coding 1 second for testing sake?

File: source/beamline/AMDetector.h [hunk 5]
```diff
@@ -276,10 +277,8 @@ int outputSize = indexStart.totalPointsTo(indexEnd);
 	/// Convenience call to access the three previous calls from a common interface.
 	bool readingND(const AMnDIndex &startIndex, const AMnDIndex &endIndex, double *outputValues) const;
 
-	// FLESH THIS ONE OUT
-	/// Returns the data from the last continuous reading in the outputValues
-	virtual bool lastContinuousReading(double *outputValues) const;
-	virtual int lastContinuousSize() const;
+	/// If the detector supports the AMDSClientDataRequest, this will return an AMDSClientDataRequest pointer for the specified period of time
+	virtual AMDSClientDataRequest* lastContinuousData(double seconds);
```
Comment (dretrex) [inline source/beamline/AMDetector.h]: `const` function.

File: source/beamline/CLS/CLSAmptekSDD123DetectorNew.h [hunk 6]
```diff
@@ -63,11 +68,12 @@ Q_OBJECT
...
-	/// Returns SingleRead as the type
-	virtual AMDetectorDefinitions::ReadMode readMode() const { return AMDetectorDefinitions::SingleRead; }
 
-	/// Returns false, because the Amptek detectors do not support continuous reads
-	virtual bool lastContinuousReading(double *outputValues) const;
+	/// Returns the type of the last acquisition
+	virtual AMDetectorDefinitions::ReadMode readMode() const { return lastReadMode_; }
+
+	/// Implemented to support returning data from the last acquire(AMDetectorDefinitions::ContinuousMode) call
+	virtual AMDSClientDataRequest* lastContinuousData(double seconds);
```
Comment (dretrex) [inline source/beamline/CLS/CLSAmptekSDD123DetectorNew.h]: `const`

**Comment by dretrex** [comment]
I know that this is just testing it from a button, but if that's all that was required to get a proof of concept going, that's awesome.  Great work on the overall data server @XuetaoLiu-CLS!

File: source/acquaman/AMAgnosticDataAPI.cpp (hunk 2 above)
Comment (davidChevrier) [inline source/acquaman/AMAgnosticDataAPI.cpp]: Definitely, my bust

File: source/actions3/actions/AMDetectorReadAction.cpp (hunk 4 above)
Comment (davidChevrier) [inline source/actions3/actions/AMDetectorReadAction.cpp]: Yep

File: source/actions3/actions/AMDetectorReadAction.cpp (hunk 3 above)
Comment (davidChevrier) [inline source/actions3/actions/AMDetectorReadAction.cpp]: Yes, not sure of a better way to handle it right now.

File: source/beamline/AMDetector.h (hunk 5 above)
Comment (davidChevrier) [inline source/beamline/AMDetector.h]: Noted.

**Pull request merged**